iroh-gateway/src/core.rs

```diff
@@ -298,87 +426,167 @@ mod tests {
+        assert_eq!(
+            format!(
+                "http://{}.ipfs.localhost:{}/world.txt",
+                test_setup.root_cid,
+                test_setup.gateway_addr.port()
+            ),
+            res.headers().get("Location").unwrap().to_str().unwrap(),
         );
-        config.set_default_headers();
 
-        let (addr, rpc_client, core_task) = spawn_gateway(Arc::new(config)).await;
+        // No trailing slash
+        let res = do_request(
+            "GET",
+            &format!("localhost:{}", test_setup.gateway_addr.port()),
+            &format!("/ipfs/{}", test_setup.root_cid),
+            None,
+        )
+        .await;
 
-        let dir = "demo";
-        let files = [
-            ("hello.txt", b"ola".to_vec()),
-            ("world.txt", b"mundo".to_vec()),
-        ];
+        assert_eq!(http::StatusCode::MOVED_PERMANENTLY, res.status());
+        assert_eq!(
+            format!(
+                "http://{}.ipfs.localhost:{}/",
+                test_setup.root_cid,
+                test_setup.gateway_addr.port()
+            ),
+            res.headers().get("Location").unwrap().to_str().unwrap()
+        );
 
-        // add a directory with two files to the store.
-        let (root_cid, _) = put_directory_with_files(&rpc_client, dir, &files).await;
+        // Trailing slash
+        let res = do_request(
+            "GET",
+            &format!("localhost:{}", test_setup.gateway_addr.port()),
+            &format!("/ipfs/{}/", test_setup.root_cid),
+            None,
+        )
+        .await;
 
-        // request the root cid as a recursive car
-        let res = {
-            let client = hyper::Client::new();
-            let uri = hyper::Uri::builder()
-                .scheme("http")
-                .authority(format!("localhost:{}", addr.port()))
-                .path_and_query(format!("/ipfs/{}/{}", root_cid, "world.txt"))
-                .build()
-                .unwrap();
-            let req = hyper::Request::builder()
-                .method("HEAD")
-                .uri(uri)
-                .body(hyper::Body::empty())
-                .unwrap();
-            client.request(req).await.unwrap()
-        };
+        assert_eq!(http::StatusCode::MOVED_PERMANENTLY, res.status());
+        assert_eq!(
+            format!(
+                "http://{}.ipfs.localhost:{}/",
+                test_setup.root_cid,
+                test_setup.gateway_addr.port()
+            ),
+            res.headers().get("Location").unwrap().to_str().unwrap()
+        );
 
-        assert_eq!(http::StatusCode::OK, res.status());
-        assert!(res.headers().get("content-length").is_some());
-        assert_eq!(res.headers().get("content-length").unwrap(), "5");
+        // IPNS
+        let res = do_request(
+            "GET",
+            &format!("localhost:{}", test_setup.gateway_addr.port()),
+            "/ipns/k51qzi5uqu5dlvj2baxnqndepeb86cbk3ng7n3i46uzyxzyqj2xjonzllnv0v8",
```

Comment:
Would be nice to add separate test for legacy notation "peerid in Base58"  that was not a valid CID (e.g., `12D3KooWJHxkQKX8C5KAyqEPhn2ssT2in4TExyG9SXxi519tycL9`)
See explainer in https://github.com/multiformats/cid-utils-website/issues/48

This is important feature of subdomain gateway: when a request is made to `example.com/ipns/id`, subdomain set up on `example.com` will return HTTP 301 redirect to proper subdomain. 
Before redirect is made, `id` is converted to `cidv1-libp2p-key-base36`.

Reply:
Done

I had to add nasty `from_peer_id` function for dealing with discrepancy between `multihash:0.16` and `multihash:0.17`. This func may be safely deleted after updateing to multihash:0.17

cc @dignifiedquire Would you mind to release multiaddr 0.17 for bringing new multihash crate to rust-libp2p?

Reply:
@vmx what't the status here?

Reply:
@dignifiedquire I'm not maintaining rust-multiaddr, that's a question for @mxinden :)

Reply:
Kicked off release process here. Please approve. https://github.com/multiformats/rust-multiaddr/pull/66